### PR TITLE
Embedded and encoded TupleMatchers

### PR DESCRIPTION
**Pull request by matt-gardner**
This PR allows for `TupleMatchers` that operate on embedded and encoded tuples, instead of just word indices.  The design is slightly complicated because we need to have access to `TextTrainer._embed_input` and `TextTrainer._get_encoder` within the `TupleMatcher`, which requires these not actually being `Layer` objects.  I haven't tested this yet; I'm setting this up so that Becky can implement an encoded tuple matcher, and I'm assuming there will be some tests associated with that.

### CODE REVIEW

File: deep_qa/layers/tuple_matchers/encoded_tuple_matcher.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        embedded_tuple2 = self.text_trainer._embed_input(tuple2)
+        tuple_encoder = self.text_trainer.get_encoder(name="tuples",
+                                                      fallback_behavior="use default encoder")
+        # We use separate encoders here in case the tuples have different shapes (e.g., different
+        # numbers of slots, or number of words per slot).
+        tuple1_encoder = EncoderWrapper(EncoderWrapper(tuple_encoder), name="tuple1_encoder")
+        tuple2_encoder = EncoderWrapper(EncoderWrapper(tuple_encoder), name="tuple2_encoder")
+        encoded_tuple1 = tuple1_encoder(embedded_tuple1)
+        encoded_tuple2 = tuple2_encoder(embedded_tuple2)
+        match_layer = TimeDistributed(TimeDistributed(TimeDistributed(self.tuple_matcher)))
```
Comment (bsharpataz) [inline deep_qa/layers/tuple_matchers/encoded_tuple_matcher.py]: This needs to be `EncoderWrapper`, not `TimeDistributed`.

File: deep_qa/layers/tuple_matchers/encoded_tuple_matcher.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                                                       "type",
+                                                       list(encoded_tuple_matchers.keys()))
+        self.tuple_matcher = encoded_tuple_matchers[tuple_matcher_choice](**tuple_matcher_params)
+
+    def __call__(self, inputs):
+        # pylint: disable=protected-access
+        tuple1, tuple2 = inputs
+        embedded_tuple1 = self.text_trainer._embed_input(tuple1)
+        embedded_tuple2 = self.text_trainer._embed_input(tuple2)
+        tuple_encoder = self.text_trainer.get_encoder(name="tuples",
```
Comment (bsharpataz) [inline deep_qa/layers/tuple_matchers/encoded_tuple_matcher.py]: `s/get_encoder/_get_encoder/`

**Comment by bsharpataz** [comment]
matt --modified the comment, want to see if it's ok?

**Comment by matt-gardner** [comment]
Sure, looks fine to me.

**Pull request merged**